test/e2e/clusterctl_upgrade.go

```diff
@@ -0,0 +1,298 @@
+			binaryURL = strings.ReplaceAll(binaryURL, "{ARCH}", runtime.GOARCH)
+		}
+
+		log.Logf("downloading clusterctl binary from %s", binaryURL)
+		binaryFile := downloadToTmpFile(binaryURL)
+		defer os.Remove(binaryFile) // clean up
+
+		err := os.Chmod(binaryFile, 0744)
+		Expect(err).ToNot(HaveOccurred(), "failed to chmod temporary file")
+
+		By("Initializing the workload cluster with older versions of providers")
+		clusterctl.InitManagementClusterAndWatchControllerLogs(context.TODO(), clusterctl.InitManagementClusterAndWatchControllerLogsInput{
+			BinaryFile:              binaryFile, // use older version of clusterctl to init the management cluster
+			ClusterProxy:            managementClusterProxy,
+			ClusterctlConfigPath:    input.ClusterctlConfigPath,
+			CoreProvider:            input.E2EConfig.GetProvidersWithOldestVersion(config.ClusterAPIProviderName)[0],
+			BootstrapProviders:      input.E2EConfig.GetProvidersWithOldestVersion(config.KubeadmBootstrapProviderName),
+			ControlPlaneProviders:   input.E2EConfig.GetProvidersWithOldestVersion(config.KubeadmControlPlaneProviderName),
+			InfrastructureProviders: input.E2EConfig.GetProvidersWithOldestVersion(input.E2EConfig.InfrastructureProviders()...),
+			LogFolder:               filepath.Join(input.ArtifactFolder, "clusters", cluster.Name),
+		}, input.E2EConfig.GetIntervals(specName, "wait-controllers")...)
+
+		By("THE MANAGEMENT CLUSTER WITH THE OLDER VERSION OF PROVIDERS IS UP&RUNNING!")
+
+		Byf("Creating a namespace for hosting the %s test workload cluster", specName)
+		testNamespace, testCancelWatches = framework.CreateNamespaceAndWatchEvents(context.TODO(), framework.CreateNamespaceAndWatchEventsInput{
+			Creator:   managementClusterProxy.GetClient(),
+			ClientSet: managementClusterProxy.GetClientSet(),
+			Name:      specName,
+			LogFolder: filepath.Join(input.ArtifactFolder, "clusters", "bootstrap"),
+		})
+
+		By("Creating a test workload cluster")
+
+		// NOTE: This workload cluster is used to check the old management cluster works fine.
+		// In this case ApplyClusterTemplateAndWait can't be used because this helpers is linked to last version of the API;
+		// so the are getting a template using the downloaded version of clusterctl, applying it, and wait for machines to be provisioned.
+
+		clusterName = fmt.Sprintf("%s-%s", specName, util.RandomString(6))
+		kubernetesVersion := input.E2EConfig.GetVariable(KubernetesVersion)
+		controlPlaneMachineCount := pointer.Int64Ptr(1)
+		workerMachineCount := pointer.Int64Ptr(1)
+
+		log.Logf("Creating the workload cluster with name %q using the %q template (Kubernetes %s, %d control-plane machines, %d worker machines)",
+			clusterName, "(default)", kubernetesVersion, controlPlaneMachineCount, workerMachineCount)
+
+		log.Logf("Getting the cluster template yaml")
+		workloadClusterTemplate := clusterctl.ConfigClusterWithBinary(ctx, binaryFile, clusterctl.ConfigClusterInput{
```

Comment:
binaryFile was optional a few lines above during init, doesn't seem to be optional here, but I assume that's WIP

Reply:
binaryFile is required both here and above because it is required to use clusterctl v1alpha3 to operate on v1alpha3 clusters

Reply:
I thought that it might be optional because we're checking via `if input.E2EConfig.HasVariable("INIT_WITH_BINARY") {` before the download and if it's mandatory we should probably just fail hard somewhere (e.g. in `BeforeEach`) and then assume the variable is set.

Reply:
I have mad this mandatory